src/grammar/block.rs

```diff
@@ -0,0 +1,415 @@
+// SPDX-License-Identifier: SEE LICENSE IN LICENSE.md
+// SPDX-FileCopyrightText: 2023 CYBERTEC PostgreSQL International GmbH
+// <office@cybertec.at>
+
+//! Implements parsing of SQL blocks from a token tree.
+
+use crate::grammar::{
+    expr, opt_expr, opt_function_invocation, opt_insert, opt_query, parse_datatype, parse_ident,
+};
+use crate::lexer::TokenKind;
+use crate::parser::Parser;
+use crate::syntax::SyntaxKind;
+use crate::ParseError;
+
+/// Looks ahead and parses a block if applicable
+pub(crate) fn opt_block(p: &mut Parser) -> bool {
+    if p.at(TokenKind::DeclareKw) || p.at(TokenKind::BeginKw) {
+        parse_block(p);
+        return true;
+    }
+    false
+}
+
+/// Parses a complete block.
+pub fn parse_block(p: &mut Parser) {
+    p.start(SyntaxKind::Block);
+    parse_declare_section(p);
+
+    p.expect(TokenKind::BeginKw);
+
+    while !p.at(TokenKind::EndKw) {
+        parse_stmt(p);
+    }
+
+    p.expect(TokenKind::EndKw);
+    parse_ident(p, 0..1);
+    p.expect(TokenKind::SemiColon);
+
+    p.finish();
+}
+
+/// <https://docs.oracle.com/cd/B28359_01/appdev.111/b28370/block.htm#CJAIABJJ>
+fn parse_declare_section(p: &mut Parser) {
+    if p.at(TokenKind::BeginKw) {
+        return;
+    }
+
+    p.start(SyntaxKind::DeclareSection);
+    p.eat(TokenKind::DeclareKw);
+    while !p.at(TokenKind::BeginKw) && !p.at(TokenKind::Eof) {
+        p.start(SyntaxKind::VariableDecl);
+
+        if !p.expect_one_of(&[TokenKind::UnquotedIdent, TokenKind::QuotedIdent]) {
+            break;
+        }
+
+        while !p.at(TokenKind::SemiColon) && !p.at(TokenKind::Eof) {
+            parse_datatype(p);
+        }
+
+        p.finish();
+
+        if !p.eat(TokenKind::SemiColon) {
+            break;
+        }
+    }
+    p.finish();
+}
+
+fn parse_stmt(p: &mut Parser) {
+    p.start(SyntaxKind::BlockStatement);
+
+    if !(opt_block(p)
+        || opt_if_stmt(p)
+        || opt_null_stmt(p)
+        || opt_return_stmt(p)
+        || opt_insert(p)
+        || opt_query(p, true)
+        || opt_procedure_call(p))
+    {
+        p.error(ParseError::ExpectedStatement);
```

Comment:
It might be already in the generated AST output, my suggestion is to pass in the current token here to learn why it failed to parse a statement.

Reply:
Hm, after giving this function a bit more thought, personally I would prefer a `match` statement iterating over the current token & forward to the the associated parse functions then. For example:

```rust
fn parse_stmt(p: &mut Parser) {
    p.start(SyntaxKind::BlockStatement);

    match p.current() {
        TokenKind::DeclareKw | TokenKind::BeginKw => { opt_block(p); },
        TokenKind::IfKw => { opt_if_stmt(p); },
        // ...
        current_token => {
            p.error(ParseError::ExpectedStatement(current_token));
            p.bump_any();
        }
    }
    p.finish();
}
```

Main reason is all these `opt_` prefixed functions check their expected starting tokens before parsing happens, which require to check the return value in order to know something was parsed or not. They become slightly simpler to follow. Another advantage is, these functions can then be better tested on their own.

I would suggest to rename the `opt_` functions to use the `parse_` prefix again, to also be consistent with existing functions for other areas.

Reply:
I agree with you, the reason I opted to introduce `opt_` functions everywhere is due to the first one being  the [`opt_function_invocation`](https://github.com/cybertec-postgresql/poc-plpgsql-analyzer/blob/main/src/grammar/function_invocation.rs#L14-L22), which is decisively more complex with it's look-ahead requirements (here it is only used for `opt_procedure_call`).

I will refactor it to more closely match your suggestion.